File: Week002/mwynholds/goooal.rb
class String
  def goooal!()
    _goooal self, ''
  end

  def _goooal(str, memo)
    return memo if str.empty?
    if str =~ /^\(\)/
      _goooal str[2..-1], (memo + 'o')
    elsif str[0] =~ /[a-zA-Z]/
      _goooal str[1..-1], (memo + str[0])
    else
      _goooal str[1..-1], memo
    end
  end
end

def assert(value)
  print ( value ? '.' : 'F' )
end

assert 'g()()()()()()()("al")'.goooal! == 'goooooooal'
assert 'g()("al")'.goooal! == 'goal'
assert 'g(’al’)'.goooal! == 'gal'

puts ''

File: Week002/mwynholds/carbonfive/Goooal.java
package carbonfive;

public class Goooal {
  public static void main(String[] args) {
    validate( g("al").equals("gal") );
    validate( g().o("al").equals("goal") );
    System.out.println();
  }

  private static void validate(boolean b) {
    System.out.print(b ? '.' : 'F');
  }

  private static String g(String s) {
    return "g" + s;
  }

  private static G g() {
    G g = (memo) -> {
      return memo;
    };
    return g;
  }

  private interface G {
    public abstract G o(String memo);
  }
}
